docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+
+
+### API Changes
+
+#### Machine
+
+> TODO: we will add this later, after we get feedback from the first daft
+> >
+> Requirements:
+>
+> * Can store the update plan: array of strings
+> * Can indicate a Machine is up to date or not and if it's on-going an update.
+
+#### MachineDeployment
+
+> TODO: we will add this later, after we get feedback from the first daft
+> >
+> Requirements:
+>
+> * Can configure external as the rollout strategy
+> * Can configure rolling as the fallback strategy when external strategy is selected
+
+#### KCP
+
+> TODO: we will add this later, after we get feedback from the first daft
+> >
+> Requirements:
+>
+> * Can configure external as the rollout strategy
+> * Can configure rolling as the fallback strategy when external strategy is selected
+
+#### External Update RuntimeExtension
+
+> TODO: we will add this later, after we get feedback from the first daft
+
+##### `CanUpdateMachine` endpoint
+##### Request
+> Requirements:
+> * Current Machine Spec
+> * Desired Machine Spec, including bootstrap and and infra machine templates
+
+##### Response
+> Requirements:
+> * Set of supported changes, probably and array of strings (the path in the object)
+> * Error
+
+##### `UpdateMachine` endpoint
+##### Request
+> Requirements:
+> * Machine reference
+
+##### Response
+> Requirements:
+> * Result: [Success/Error/InProgress]
+> * Retry in X seconds
+
+### Security Model
+
+On the core CAPI side, the security model for this feature is very straight forward: CAPI controllers only require to read/create/update CAPI resources and those controllers are the only ones that need to modify the CAPI resources. Moreover, the controllers that need to perform these actions already have the necessary permissions over the resources they need to modify.
+
+However, each external updater should define their own security model. Depending on the mechanism used to update machines in-place, different privileges might be needed, from scheduling privileged pods to SSH access to the hosts. Moreover, external updaters  might need RBAC to read CAPI resources.
+
+### Risks and Mitigations
+
+1. One of the risks for this process could be that during a single node cluster in-place update, extension implementation might decline the update and that would result in falling back to rolling update strategy, which could lead to breaking a cluster if not using the feature as intended (allowing fallback strategy for this specific case). Good documentation with examples can be provided to avoid such scenarios.
+
+## Additional Details
+
+### Test Plan
+
+To test the external update strategy, we will implement a "CAPD Kubeadm Updater." This will serve as a reference implementation and will be integrated into CAPI CI. In-place updates will be performed by executing a set of commands in the container, similar to how it is currently implemented for cloud config when machine is bootstrapped.
```

Comment:
```suggestion
To test the external update strategy, we will implement a "CAPD Kubeadm Updater". This will serve as a reference implementation and will be integrated into CAPI CI. In-place updates will be performed by executing a set of commands in the container, similar to how it is currently implemented for cloud config when machine is bootstrapped.
```